When getting a connection, last_used_time is checked. If its more than…

kxk302 opened the pull request:
… a specific value (configurable), connection is considered stale, it is dropped and a new connection is created. This is to address the issue with dropped stale connections.

trel commented:
Hi Kaivan, thanks - yes, Codacy has some opinions sometimes - I think this is fine.  We haven't learned how to bend Codacy to our will quite yet - it's running its default set of rules.

This looks really nice.   Tests!

Can you tell us a bit more about the scenario / use case you have where this is necessary?

d-w-moore commented:
Kaivan, just to add my input to Terrell's (as I'll often be involved integrating and testing fixes for the Python client), what is the definition of a stale connection?  Does the staleness relate to an expiry of a PAM password, or just a typical socket timeout ? And does a stale connection usually throw an exception if you attempt to use it?

kxk302 commented:
Thank you guys!

To provide more context, our Galaxy software in the Test environment is using irods for storage. Upon startup, Galaxy would create an irods session, which it would maintain. Test calls to irods were fine. Afterwards, when there was a period of inactivity, it seemed, the connections would get dropped. Afterwards, the first call to irods would fail with an exception, but subsequent calls would be fine. 

I posted in irods forum and one suggestion was to create a session for each irods calls and then drop it. I implemented this and the exceptions disappeared, but now we were paying a performance penalty, that would be noticeable when we were loading thousands of file that were part of a collection.

So, we decided to come up with a hybrid approach. We create an irods session in Galaxy and maintain it. However, if the connection was not used for a certain amount of time (we need to find the value with some experimentation), the connection would proactively be dropped and re-created (we thought it makes sense to implement this change in irods client code, not in Galaxy). This way, we have the desired performance and can also solve the dropped connection problem we had before.

I need to test this in Galaxy still (which would be an integration test), but we are hopeful/confident this willsolve the problem.

Let me know if you need more info/clarification.

Best,
-Kaivan

kxk302 commented:
Here is the exception we were getting: 

timeout: timed out
  File "irods/connection.py", line 86, in recv
    msg = iRODSMessage.recv(self.socket)
  File "irods/message/__init__.py", line 74, in recv
    rsp_header_size = _recv_message_in_len(sock, 4)
  File "irods/message/__init__.py", line 27, in _recv_message_in_len
    buf = sock.recv(size_left, socket.MSG_WAITALL)
NetworkException: Could not receive server response
  File "galaxy/web/framework/decorators.py", line 295, in decorator
    rval = func(self, trans, *args, **kwargs)
  File "galaxy/webapps/galaxy/api/tools.py", line 459, in create
    return self._create(trans, payload, **kwd)
  File "galaxy/webapps/galaxy/api/tools.py", line 531, in _create
    output_dict = output.to_dict()
  File "galaxy/model/__init__.py", line 3197, in to_dict
    file_size=int(hda.get_size()),
  File "galaxy/model/__init__.py", line 2594, in get_size
    return self.dataset.get_size()
  File "galaxy/model/__init__.py", line 2294, in get_size
    return self._calculate_size()
  File "galaxy/model/__init__.py", line 2281, in _calculate_size
    return self.object_store.size(self)
  File "galaxy/objectstore/__init__.py", line 283, in size
    return self._invoke('size', obj, **kwargs)
  File "galaxy/objectstore/__init__.py", line 271, in _invoke
    return self.__getattribute__("_" + delegate)(obj=obj, **kwargs)
  File "galaxy/objectstore/__init__.py", line 640, in _size
    return self._call_method('_size', obj, 0, False, **kwargs)
  File "galaxy/objectstore/__init__.py", line 850, in _call_method
    object_store_id = self.__get_store_id_for(obj, **kwargs)
  File "galaxy/objectstore/__init__.py", line 870, in __get_store_id_for
    if store.exists(obj, **kwargs):
  File "galaxy/objectstore/__init__.py", line 274, in exists
    return self._invoke('exists', obj, **kwargs)
  File "galaxy/objectstore/__init__.py", line 271, in _invoke
    return self.__getattribute__("_" + delegate)(obj=obj, **kwargs)
  File "galaxy/objectstore/irods.py", line 431, in _exists
    in_irods = self._data_object_exists(rel_path)
  File "galaxy/objectstore/irods.py", line 308, in _data_object_exists
    self.session.data_objects.get(data_object_path)
  File "irods/manager/data_object_manager.py", line 45, in get
    parent = self.sess.collections.get(irods_dirname(path))
  File "irods/manager/collection_manager.py", line 17, in get
    result = query.one()
  File "irods/query.py", line 220, in one
    results = self.execute()
  File "irods/query.py", line 174, in execute
    result_message = conn.recv()
  File "irods/connection.py", line 90, in recv
    raise NetworkException("Could not receive server response")

d-w-moore commented:
Kaivan, 
thank you,  I shall experiment.  Do you think it would help to have the default behavior simply test connections as they're brought out of the idle pool for service, to see if they are stale?

kxk302 commented:
I revised the code a bit based on a self review. 

@d-w-moore Thats a good idea (and an optimization, as there is no need to check the last_used_time of a new connection). I can move the refresh code into the try clause in get_connection()
            try:
                conn = self.idle.pop()
               // Refresh logic moves here 
            except KeyError:
                conn = Connection(self, self.account)

trel commented on irods/test/pool_test.py:
```diff
@@ -60,24 +60,70 @@ def test_release_disconnected(self):
+        # Release the connection (goes from active to idle queue)
+        # Again, get a connection. Should get the same connection back.
+        # I.e., the object IDs should match. However, the new connection
+        # should have a more recent 'last_iused_time'
```
last_iused_time -> last_used_time

kxk302 replied:
Fixed locally. Thanks!

kxk302 commented:
Pushed my change with a test-data folder.

d-w-moore commented:
Seems fine except that 7 of the new tests are failing (at least , for me, on Python 2.7 ). Other than that, the whole test suite passes at bench, which is good.
[tests_kxk.txt](https://github.com/irods/python-irodsclient/files/5385223/tests_kxk.txt)

kxk302 commented:
> Seems fine except that 7 of the new tests are failing (at least , for me, on Python 2.7 ). Other than that, the whole test suite passes at bench, which is good.
> [tests_kxk.txt](https://github.com/irods/python-irodsclient/files/5385223/tests_kxk.txt)

I see. So, python_irodsclient code is Python2 complaint? That would explain 4 of the errors (FileNotFoundError is a Python3 exception). I will get the tests to pass with Python2.7 and then push the changes. Thanks

kxk302 commented:
Ok, the fix was rather easy. I changed FileNotFoundError to IOError to make the code Python2 compliant. Also, fixed the path to irods environment files in pool_test.py. All tests pass on my machine using Python 2.7

trel commented:
Yes, Python 2.7 or 3.4+.

If we can confirm things work in 3.4+, then I think we've got a winner.

kxk302 commented:
Just verified, pool_test.py tests pass with both Python 2.7 and 3.7.

trel commented:
Excellent.  Thanks!

d-w-moore commented:
@kxk302 , We're getting there.... But I'm still getting some failures in the new tests , on that latest  [commit](https://github.com/kxk302/python-irodsclient/commit/45640c87a94841576b32e11cd64fac48ab20c6d1). 

This is on Ubuntu 18.04,  python3.6.9 , within a virtualenv.

I'll attempt it next without the virtual environment , to see if that is somehow the issue with my failed instances.

```
(py369) irods@a036bfacfc97:~/python-irodsclient/irods/test$ python runner.py 

Running tests...
----------------------------------------------------------------------
  test_list_acl (access_test.TestAccess) ... OK (0.106s)
  test_raw_acls__207 (access_test.TestAccess) ... OK (0.174s)
     (... all
            passing ...)
  test_user_dn (user_group_test.TestUserGroup) ... OK (0.069s)
  test_user_metadata (user_group_test.TestUserGroup) ... OK (0.081s)

======================================================================
ERROR [0.036s]: test_connection_last_used_time (pool_test.TestPool)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/lib/irods/python-irodsclient/irods/test/pool_test.py", line 92, in test_connection_last_used_time
    self.assertTrue(last_used_time_2 > last_used_time_1)
AssertionError: False is not true

======================================================================
ERROR [1.033s]: test_no_refresh_connection (pool_test.TestPool)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/lib/irods/python-irodsclient/irods/test/pool_test.py", line 180, in test_no_refresh_connection
    self.assertTrue(last_used_time_2 > last_used_time_1)
AssertionError: False is not true

======================================================================
ERROR [5.038s]: test_refresh_connection (pool_test.TestPool)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/lib/irods/python-irodsclient/irods/test/pool_test.py", line 137, in test_refresh_connection
    self.assertNotEqual(conn_obj_id_1, conn_obj_id_2)
AssertionError: 140391518491704 == 140391518491704

----------------------------------------------------------------------
Ran 151 tests in 419.186s

FAILED (errors=3, skipped=13)

Generating XML reports...
(py369) irods@a036bfacfc97:~/python-irodsclient/irods/test$ 
```

kxk302 commented:
> @kxk302 , We're getting there.... But I'm still getting some failures in the new tests , on that latest [commit](https://github.com/kxk302/python-irodsclient/commit/45640c87a94841576b32e11cd64fac48ab20c6d1).
> 
> This is on Ubuntu 18.04, python3.6.9 , within a virtualenv.
> 
> I'll attempt it next without the virtual environment , to see if that is somehow the issue with my failed instances.
> 
> ```
> (py369) irods@a036bfacfc97:~/python-irodsclient/irods/test$ python runner.py 
> 
> Running tests...
> ----------------------------------------------------------------------
>   test_list_acl (access_test.TestAccess) ... OK (0.106s)
>   test_raw_acls__207 (access_test.TestAccess) ... OK (0.174s)
>      (... all
>             passing ...)
>   test_user_dn (user_group_test.TestUserGroup) ... OK (0.069s)
>   test_user_metadata (user_group_test.TestUserGroup) ... OK (0.081s)
> 
> ======================================================================
> ERROR [0.036s]: test_connection_last_used_time (pool_test.TestPool)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/var/lib/irods/python-irodsclient/irods/test/pool_test.py", line 92, in test_connection_last_used_time
>     self.assertTrue(last_used_time_2 > last_used_time_1)
> AssertionError: False is not true
> 
> ======================================================================
> ERROR [1.033s]: test_no_refresh_connection (pool_test.TestPool)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/var/lib/irods/python-irodsclient/irods/test/pool_test.py", line 180, in test_no_refresh_connection
>     self.assertTrue(last_used_time_2 > last_used_time_1)
> AssertionError: False is not true
> 
> ======================================================================
> ERROR [5.038s]: test_refresh_connection (pool_test.TestPool)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/var/lib/irods/python-irodsclient/irods/test/pool_test.py", line 137, in test_refresh_connection
>     self.assertNotEqual(conn_obj_id_1, conn_obj_id_2)
> AssertionError: 140391518491704 == 140391518491704
> 
> ----------------------------------------------------------------------
> Ran 151 tests in 419.186s
> 
> FAILED (errors=3, skipped=13)
> 
> Generating XML reports...
> (py369) irods@a036bfacfc97:~/python-irodsclient/irods/test$ 
> ```

Hmm, I run the tests on Mac, in a virtual environment with Python 3.7 and don't have any errors. Would you like to try Python3.7 and see if that makes a difference?

d-w-moore commented:
> Hmm, I run the tests on Mac, in a virtual environment with Python 3.7 and don't have any errors. Would you like to try Python3.7 and see if that makes a difference?

I'll compile 3.7 from source and attempt it.

d-w-moore commented:
Well, @kxk302 , bad news is that under 3.7  I still get the same three errors from my last comment above, located at
   - "/var/lib/irods/python-irodsclient/irods/test/pool_test.py", line 92
   - "/var/lib/irods/python-irodsclient/irods/test/pool_test.py", line 180
   - "/var/lib/irods/python-irodsclient/irods/test/pool_test.py", line 137

The good news is running tests on 3.7 have exposed another failure in my code I didn't know about! : )

Now , I guess, to figure out what could be the difference in our environments, (other than yours being Mac and mine Linux, which shouldn't actually matter....)

kxk302 commented:
> Well, @kxk302 , bad news is that under 3.7 I still get the same three errors from my last comment above, located at
> 
> * "/var/lib/irods/python-irodsclient/irods/test/pool_test.py", line 92
> * "/var/lib/irods/python-irodsclient/irods/test/pool_test.py", line 180
> * "/var/lib/irods/python-irodsclient/irods/test/pool_test.py", line 137
> 
> The good news is running tests on 3.7 have exposed another failure in my code I didn't know about! : )
> 
> Now , I guess, to figure out what could be the difference in our environments, (other than yours being Mac and mine Linux, which shouldn't actually matter....)

If you like, we could have a brief Zoom meeting to look at this issue. I'm in the middle something now, but should be free after 2:00 pm.

d-w-moore commented:
> 
> If you like, we could have a brief Zoom meeting to look at this issue. I'm in the middle something now, but should be free after 2:00 pm.

That would work ! Sending you an email so that you have my address!

kxk302 commented:
> > If you like, we could have a brief Zoom meeting to look at this issue. I'm in the middle something now, but should be free after 2:00 pm.
> 
> That would work ! Sending you an email so that you have my address!

@d-w-moore You can email me at kxk302@psu.edu.

d-w-moore commented:
@trel I've verified this runs with all tests passing on Python 3.4+ and 2.7 ... Centos and Ubuntu18.
That's when tested in combination with the last two commits on this PR/branch:  #219

d-w-moore commented:
@kxk302 everything checks out good in tests on my side, just need you to squash it all down to a single commit on your Github branch... choose a commit message that is descriptive of the problem, too, since I don't believe an actual code issue was filed!

kxk302 commented:
@d-w-moore I squashed all commits into one.

kxk302 commented:
@d-w-moore question: when this gets merged, would there be a new version of python-irodsclient (with this change) that I can refer to in my Galaxy dependency file? Thx

trel commented:
Yes, we can release 0.8.4 pretty soon after this.

kxk302 commented:
> Yes, we can release 0.8.4 pretty soon after this.

Great, Thanks!

d-w-moore commented:
@kxk302,  I've just absorbed your squashed commit in with my PR to fix some other Python3 test snafu's ...
(Credit is still yours on the commit containing your changes, of course :) )
I'll mail you with the new ref when it's merged. And per @trel's comment just now (thanks live page update!) I see we'll have a new version soon!

kxk302 commented:
Thanks @trel! So, what is the process for generating the artifact for the new version? Is there a way for me to see the progress/copletion?

trel commented:
For the official release to pypi from the master branch, it's just me running...

```
python setup.py bdist_wheel --universal
twine check dist/*
twine upload dist/*
```

I need to get the CHANGELOG and version tweaked, but that can happen today/tomorrow.
